### ISSUE TITLE
toctree does not consider "only" directive

### ISSUE THREAD

**Issue by ngzhian** [sphinx-doc/sphinx#9819]
### Describe the bug

`toctree` when used in templates does not consider `.. only` directive.

In `index.rst` Use `.. only` to only add some documents for singlehtml builder.

In the template, using `toctree`, and building multi-page html, all the toc tree shows up.

<img width="120" alt="Screen Shot 2021-11-04 at 4 57 51 PM" src="https://user-images.githubusercontent.com/1749303/140436359-e47a24ce-76ba-4ec7-bf21-69101e09a797.png">

Since this is a multi-page html build, should the `.. only` cause the `toctree` to only have `two`?

### How to Reproduce

Use sphinx-quickstart to set up a project with any values.

Update source/index.rst with this toc:
```
.. only:: singlehtml_builder

   .. toctree::
      one

.. toctree::
   :maxdepth: 2
   :caption: Contents:

   two
```
Create `source/one.rst` and `source/two.rst`.
Create a custom `_templates/layout.html` that uses `toctree()`:
```
{% extends "!layout.html" %}

{%- block sidebar1 %}
{{ super() }}
{{ toctree() }}
{% endblock %
```

`sphinx-build html`

Open `build/html/index.html`

### Expected behavior

TOC only contains link to "two"

### Your project

https://github.com/ngzhian/sphinx-repro

### OS

Linux

### Python version

3.9.7

### Sphinx version

4.2.0

**Comment by tk0miya** [comment]
Yes. The "only" directive has many structural problems. This is one of them. The reason for the problems is the content filtering feature conflicts with the incremental build feature. So we gave up to support the full features. Officially, it is recommended to use to filter paragraphs.

**Comment by ngzhian** [comment]
Thanks for your explanation. I have a PR up at https://github.com/sphinx-doc/sphinx/pull/9830, is there interest to fix this "structural problem" for toctree? If yes, I can fix up that PR (now there is a merge conflict).

**Comment by tk0miya** [comment]
I consider #9830 can resolve one of the structural problems. For example, it does not fix the "next page" and "previous page" links. They sometimes point to filtered documents.

To resolve the whole of them, we have to disable the incremental build feature if users use the "only" directive. So we should choose another way if we'll resolve it.

**Comment by lemketron** [comment]
> Thanks for your explanation. I have a PR up at #9830, is there interest to fix this "structural problem" for toctree? If yes, I can fix up that PR (now there is a merge conflict).

@ngzhian It looks like #9830 was closed (but wasn't merged)? 

Is there any other way to conditionally include references inside a toctree?

I'd like to be able to do something like this (to conditionally certain pages) but can't figure out how:
```
.. toctree::
   :maxdepth: 2
   :caption: Contents

   FIRST  <first.rst>
   SECOND <second.rst>

   .. only:: internal

      INTERNAL NOTES <secret-stuff.rst>
```

This following does manage to hide the referenced INTERNAL pages from the TOC, but the "INTERNAL SECRETS" heading (caption) still appears even in the non-internal build:
```
.. only:: internal

   .. toctree::
      :maxdepth: 2
      :caption: INTERNAL SECRETS
   
      INTERNAL NOTES <secret-stuff.rst>
```

**Comment by ngzhian** [comment]
my PR wasn't complete and didn't fix all use cases, I did not continue
working on it

**Comment by AdrianVollmer** [comment]
>  Officially, it is recommended to use to filter paragraphs.

What do you mean by "filter paragraphs"?
